From 985df02d8e6301add231e2989b357af4fc6d7c03 Mon Sep 17 00:00:00 2001
From: David Pennington <dpenning@chromium.org>
Date: Wed, 04 Feb 2026 09:54:43 -0800
Subject: [PATCH] [Vertical Tabs] Protect against updates during animating out a group

There are cases where the collection node is set to nullptr, however
theres a bunch of assumptions in VTGView that it always exists. these
issues are fixed by early returns or CHECKing based on the function.

Fixed: 481440222
Change-Id: I62b634494158b0ddbcbbf808718e9c9830ea1e9c
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7543160
Commit-Queue: David Pennington <dpenning@chromium.org>
Reviewed-by: Caroline Rising <corising@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1579555}
---

--- a/chrome/browser/ui/views/tabs/vertical/vertical_tab_group_view.cc
+++ b/chrome/browser/ui/views/tabs/vertical/vertical_tab_group_view.cc
@@ -46,6 +46,7 @@ constexpr int kGroupHeaderVerticalMargin
 constexpr int kTabLeadingPadding = 10;
 
 const TabGroup* GetTabGroupFromNode(TabCollectionNode* node) {
+  CHECK(node);
   return static_cast<const tabs::TabGroupTabCollection*>(
              std::get<const tabs::TabCollection*>(node->GetNodeData()))
       ->GetTabGroup();
@@ -178,6 +179,11 @@ views::ProposedLayout VerticalTabGroupVi
 
 void VerticalTabGroupView::ToggleCollapsedState(
     ToggleTabGroupCollapsedStateOrigin origin) {
+  // If the group is in the process of being closed, then ignore updates.
+  if (!collection_node_) {
+    return;
+  }
+
   collection_node_->GetController()->ToggleTabGroupCollapsedState(
       GetTabGroupFromNode(collection_node_), origin);
 }
@@ -204,10 +210,17 @@ bool VerticalTabGroupView::IsViewDraggin
 }
 
 void VerticalTabGroupView::ResetCollectionNode() {
+  node_destroyed_subscription_ = {};
+  data_changed_subscription_ = {};
   collection_node_ = nullptr;
 }
 
 void VerticalTabGroupView::OnDataChanged() {
+  // If the group is in the process of being closed, then ignore updates.
+  if (!collection_node_) {
+    return;
+  }
+
   tab_group_visual_data_ =
       *GetTabGroupFromNode(collection_node_)->visual_data();
   group_header_->OnDataChanged(&tab_group_visual_data_);
